### DISCUSSION TITLE
Multiple installations on the same server

### DISCUSSION THREAD

**Discussion by vittore** [netbox-community/netbox#9679] [Q&A]
Hi, 

we have some schools, every school has its autonomies, infrastructures, administrators, organizations, devices,... and so on.
We want to put multiple netbox installations on the same server, using the same python files, but using different databases.
So, for example, school 1 access to URL https://netbox.ourdomain/school1 and schoool2 to   URL https://netbox.ourdomain/school2 or (if it's simple) https://netbox-school1.ourdomain and https://netbox-school2.ourdomain

Is it possible? How to do this?

Thanks in advance for your ideas.
v.

**Comment by kkthxbye-code** [comment]
I think you should be able to just have different configuration.py files and using the [NETBOX_CONFIGURATION](https://github.com/netbox-community/netbox/blob/c1b7f09530f0293d0f053b8930539b1d174cd03b/docs/configuration/index.md) environment variable.

I would always recommend containerization instead though.

**Comment by mtinberg** [comment]
You'd want to have different directories for uploaded static files, and they may end up wanting their own difference Scripts and Reports too, which can be different paths in configuration.py.  I'm not sure whether it is more cumbersome to manage separate configuration.py files and directories where any per-instance state is located, or just eating the 400MB/per-instance space usage to keep each virtualhost totally isolated with its own virtualenv.  As kthxbye said, containers are a way to manage this, by running multiple instances with the same base image and different configs and storage backends, but if you don't have container servers set up that may just be adding another layer to manage to the problem you need to solve.

Regardless of whether you set them up totally separately, with multiple databases you are already in the position of needing to run the upgrade script separately for each instance, whether you keep the maintenance totally separate, so you can have different maintenance windows for each instance if you need, is up to you.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Comment by vittore** [comment]
Yes, containers could be a solution but they insert a CPU impact: with N schools I get N PostgreSQL, N unicorn, N redis,... 
Using standard installation with N schools I get 1 PostgreSQL with N databases, 1 unicorn with N sites, 1 redis,...
So I'm trying to explore better ```NETBOX_CONFIGURATION``` functionality.

v.

**Reply by candlerb** [reply]
> Yes, containers could be a solution but they insert a CPU impact: with N schools I get N PostgreSQL, N unicorn, N redis,...

That's not true. You can have N Netbox containers connecting to a single Postgres container (with N databases) and the same Redis container (using different DB IDs).

I'm not aware of any way to run a single gunicorn serving multiple Netbox instances simultaneously. They would each have to have a different global `netbox.settings` object, which AFAICS means they must be separate Python interpreter processes.

In principle, you can share the Apache/Nginx front-end between multiple gunicorns, with separate vhosts for each Netbox.  However if they are containers, the awkward bit is what to do with `/static/`.  If you use lxd containers, then you can frig the proxy to dig directly into the container filesystem:

```
	Alias /static /var/lib/lxd/containers/netbox/rootfs/opt/netbox/netbox/static/
	<Directory /var/lib/lxd/containers/netbox/rootfs/opt/netbox/netbox/static/>
		Options Indexes FollowSymLinks MultiViews
		AllowOverride None
		Require all granted
	</Directory>

	<Location /static>
		ProxyPass !
	</Location>

        ProxyPreserveHost On
        RequestHeader set X-Forwarded-Proto "https"
	ProxyPass / http://10.112.76.119:8001/
	ProxyPassReverse / http://10.112.76.119:8001/
```

Or if you are running N instances of exactly the same version of Netbox, then you can just make one local copy of the `/static/` directory for the proxy.  Anyway: the overhead of an extra Apache proxy per Netbox is pretty tiny, compared with Netbox itself.

**Comment by RuanM-CW** [comment]
Optionally you can run a docker in docker to provide each school its own login to its own infrastructure, this is where Portainer will help you manage the infrastructure of the multiple sites through stacks.
from there deploy all the schools as docker containers with their own configuration.py this should be more manageable.

https://hub.docker.com/r/linuxserver/netbox

This way you know exactly which site you working on, you can configure WireGuard from your laptop to each site's Netbox instance through an RMM tool. :)

-----

I am sitting with a similar issue, I have the answer though, I'm just very limited in my lab environment to accomplish this right now. :/

Need to find another way if anyone is able to advise? (would be much appreciated)

**Comment by gdprdatasubect** [comment]
Seems like a XY-Problem to me, it would be far better to unify the netbox-instances and go for a common ground.
Reasons:
- Employees can easily share knowledge between schools
- Higher hierarchical levels always see all lower infrastructure, not having to deal with multiple instances when gathering information
- Common Naming-Schemes and enforcement.

It is pretty easy to limit what each individual unit can see with usergroups and permissions, e.g. Admin A sees all devices that are in sites with group A, all circuits connecting these sides etc.
We limited our managing organization to only see what they bought with their budget and nothing else :)

**Comment by vittore** [comment]
Very interesting point of view.

Thanks for sharing.
v.

Il giorno gio 27 lug 2023 alle ore 00:14 gdprdatasubect <
***@***.***> ha scritto:
